standalone-demo/rating-element.js

```diff
@@ -1,71 +1,100 @@
 class RatingElement extends HTMLElement {
+  static get observedAttributes() {
+    return ['value', 'max'];
+  }
+
   constructor() {
     super();
 
-    this.ratingElements = [];
-
-    // Handle default data sent as attributes.
-    // Make sure value and number are last since they invoke render actions.
-    this._defaultClass = this.getAttribute('default-class') || 'x-rating';
-    this._markedClass = this.getAttribute('marked-class') || 'x-marked';
-    this.value = parseInt(this.getAttribute('value')) || 0;
-    this.number = parseInt(this.getAttribute('number')) || 5;
+    this.choices = [];
+    this._classes = {
+      choice: this.getAttribute('choice-class') || 'choice',
+      checked: this.getAttribute('checked-class') || 'checked',
+      highlighted: this.getAttribute('highlighted-class') || 'highlighted'
+    };
 
-    // Initial value highlight.
-    this.highlight(this.value - 1);
+    this.onHighlight = this.onHighlight.bind(this);
+    this.onRate = this.onRate.bind(this);
+  }
 
-    // Reset hovering state to an actual state once it stops.
-    this.addEventListener('mouseout', e => this.highlight(this.value - 1));
+  onHighlight(e) {
+    if (e.type === 'mouseleave') return this.highlightChoices();
+    const index = this.choices.indexOf(e.target);
+    if (index !== -1) this.highlightChoices(index);
   }
 
-  get value() {
-    return this._value;
+  onRate(e) {
+    const index = this.choices.indexOf(e.target);
+    if (index === -1) return;
```

Comment:
Can this even happen?

Reply:
It might if someone manually adds more child elements to `x-rating-element`. This listener captures click events on all (grand)children not necessarily choices.